kaldb/src/test/java/com/slack/kaldb/clusterManager/CacheSlotAssignmentServiceTest.java

```diff
@@ -0,0 +1,1201 @@
+package com.slack.kaldb.clusterManager;
+
+import static com.slack.kaldb.config.KaldbConfig.DEFAULT_START_STOP_DURATION;
+import static org.assertj.core.api.AssertionsForClassTypes.assertThat;
+import static org.awaitility.Awaitility.await;
+import static org.junit.Assert.assertTrue;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.doCallRealMethod;
+import static org.mockito.Mockito.spy;
+
+import brave.Tracing;
+import com.google.common.util.concurrent.Futures;
+import com.slack.kaldb.metadata.cache.CacheSlotMetadata;
+import com.slack.kaldb.metadata.cache.CacheSlotMetadataStore;
+import com.slack.kaldb.metadata.replica.ReplicaMetadata;
+import com.slack.kaldb.metadata.replica.ReplicaMetadataStore;
+import com.slack.kaldb.metadata.zookeeper.MetadataStore;
+import com.slack.kaldb.metadata.zookeeper.ZookeeperMetadataStoreImpl;
+import com.slack.kaldb.proto.config.KaldbConfigs;
+import com.slack.kaldb.proto.metadata.Metadata;
+import com.slack.kaldb.testlib.MetricsUtil;
+import io.micrometer.core.instrument.MeterRegistry;
+import io.micrometer.core.instrument.simple.SimpleMeterRegistry;
+import java.io.IOException;
+import java.time.Instant;
+import java.time.temporal.ChronoUnit;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.UUID;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.stream.Collectors;
+import org.apache.curator.test.TestingServer;
+import org.assertj.core.api.Assertions;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+
+public class CacheSlotAssignmentServiceTest {
```

Comment:
Now that we have cache assignment, let's add an end to end test for the snapshot lifecycle. i.e,  indexer creates a snapshot, assigner assigns the snapshot to cache node picks it up and serves queries from it.

Reply:
Let's do this test as a separate PR, as that's a bit outside of the scope for the cache slot assignment. 

We'll first need to add some missing individual lifecycle tests prior to writing a complete end to end test. As of today we have no tests for the indexer for the complete lifecycle, from message ingestion to rollover and snapshot creation. Similarly while we have tests for the caching chunk manager, we have no cache node lifecycle tests. 

I've added tasks to the development notes for each of these tests, and then the one for the complete lifecycle from message ingestion to query nodes.